WIP Add ECPublicKey / ECPrivateKey interfaces

cipherboy opened the pull request:
WIP; more details to follow. 

Related: #226.

cipherboy commented on lib/jss.map:
```diff
@@ -394,3 +394,9 @@ Java_org_mozilla_jss_nss_Buffer_Free;
     local:
         *;
 };
+JSS_4.6.2 {
```
I've done a terrible job maintaining this map. @edewata, @jmagne, or @ladycfu -- do you know if it is possible to retroactively update the version numbers that functions were added in? :-)

cipherboy commented on org/mozilla/jss/pkcs11/PK11ECPrivateKey.java:
```diff
@@ -22,21 +27,25 @@ protected PK11ECPrivateKey(byte[] pointer) {
-//      }
-//    }
+    public ECParameterSpec getParams() {
+        try {
```
I think we can actually drop this try/catch block.

cipherboy commented on org/mozilla/jss/util/ECCurve.java:
```diff
@@ -0,0 +1,199 @@
+        return null;
+    }
+
+    // OID Base Arcs
```
@edewata how should I store these? If I introduce them in the `jss.util.EC` class, they'll be publicly accessible, which I guess is fine. Perhaps I should make a new `jss.util.ECOIDs` class for storing them? I can't use them in this enum because Java complains about forward references. I think it'd make the definition up top a lot easier and a lot less error prone. We can also deduplicate them out of `PKC11KeyPairGenerator` once we get a final resting place for them.

Thoughts?

cipherboy replied:
I've introduced an ECOIDs class which I'll use for this.

cipherboy commented:
@edewata I think this is ready for review. I'm of the opinion to add more curves when we need them and not grow the Curve database too quickly. A lot of those curves simply don't get used.

edewata commented on org/mozilla/jss/pkcs11/PK11ECPublicKey.java:
```diff
@@ -4,49 +4,37 @@
+        return EC.decodeNSSPoint(pointData);
+    }
+
+    public byte[] getWBA() {
```
Instead of adding getWBA() we could simply make getWByteArray() public.

cipherboy replied:
Ah, sure, good catch.

cipherboy replied:
I'm confused why this is actually needed. I might remove it in the future, but for now I've done this.

edewata commented on org/mozilla/jss/util/ECCurve.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    public String[] getNames() {
+        return this.names;
```
It's not necessary to use `this`.

edewata commented on org/mozilla/jss/util/ECCurve.java:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    public OBJECT_IDENTIFIER[] getOIDs() {
+        return this.oids;
```
It's not necessary to use `this`.

edewata commented on org/mozilla/jss/pkcs11/PK11ECPrivateKey.java:
```diff
@@ -19,24 +24,21 @@ protected PK11ECPrivateKey(byte[] pointer) {
-//      }
-//    }
+    public ECParameterSpec getParams() {
+        PK11PubKey publicKey = this.getPublicKey();
```
It's not necessary to use `this`.

cipherboy commented:
@edewata Ok, this is fine. I've tested this against PKI and [opened a PR](https://github.com/dogtagpki/pki/pull/243) addressing issues I found with `nssdb.py`.

After that, I was able to self-host a PKI instance with EC certs and have it work with Tomcat just fine: 

```
$./testssl.sh https://localhost.localdomain:8443

~snip~

 Testing server defaults (Server Hello) 

 TLS extensions (standard)    "renegotiation info/#65281" "extended master secret/#23"
 Session Ticket RFC 5077 hint no -- no lifetime advertised
 SSL Session ID support       yes
 Session Resumption           Tickets no, ID: no
 TLS clock skew               0 sec from localtime
 Signature Algorithm          ECDSA with SHA512
 Server key size              EC 521 bits
 Server key usage             Digital Signature, Data Encipherment, Key Agreement
 Server extended key usage    TLS Web Server Authentication
 Serial / Fingerprints        03 / SHA1 AE46DF88ED2D93E539BB90665B4DE32EB1ABAB1B
                              SHA256 7F7BF4E7B196CC8A7B98189CF3710FCD97298732364B151578EED931F83A60EC
 Common Name (CN)             localhost.localdomain
 subjectAltName (SAN)         localhost.localdomain 
 Issuer                       CA Signing Certificate (EXAMPLE)
 Trust (hostname)             Ok via SAN and CN (same w/o SNI)
 Chain of trust               NOT ok (self signed CA in chain)
 EV cert (experimental)       no 
 ETS/"eTLS", visibility info  not present
 Certificate Validity (UTC)   719 >= 60 days (2019-08-22 20:32 --> 2021-08-11 20:32)
 # of certificates provided   2
 Certificate Revocation List  --
 OCSP URI                     http://localhost.localdomain:8080/ca/ocsp
 OCSP stapling                not offered
 OCSP must staple extension   --
 DNS CAA RR (experimental)    not offered
 Certificate Transparency     --
```

Is there anything else I need to test or can I go ahead and merge this?
